iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -16,45 +17,164 @@
 
 #include "iceoryx_posh/internal/roudi/service_registry.hpp"
 
+#include <algorithm>
+
 namespace iox
 {
 namespace roudi
 {
-void ServiceRegistry::add(const CaproIdString_t& service, const CaproIdString_t& instance)
+cxx::expected<ServiceRegistry::ServiceRegistryError>
+ServiceRegistry::add(const capro::ServiceDescription& serviceDescription)
 {
-    cxx::set::add(m_serviceMap[service].instanceSet, instance);
+    // Forbid duplicate service descriptions entries
```

Comment:
Please use the `cxx::set` again or a `std::set` and replace it later with an optimized implementation (the `set` is one part of the `map` implementation).

Reply:
A set is not part of a map implementation. Our current `cxx::set` abstracts away the duplicate removal but not in an efficient way. Still, I already discussed it with @mossmaurice and what we need for fast removal and insertion is a structure with 1) random access and 2) efficient insertion and removal. A set cannot help us with 1) in the long term, we need a vector with some addons, quite similar to the `Objectpool`. But the Objectpool can be improved to allow O(1) insertion. 

Long story short we can use the `cxx::set` abstraction for now but need some thin wrapper around a vector with properties 1) and 2) later.

Reply:
I don't see why we should use our `cxx::set` wrapper here. To me it seems like we need a bit more advanced but still light-weight data structure than the `cxx::vector`. @elfenpiff Can you elaborate?

Reply:
To my knowledge `cxx::set` is a wrapper around a vector which takes care of duplicate elimination, albeit in an inefficient way (linear search). If we can use this instead of implementing duplicate elimination here, we should.

Reply:
Let's resolve this topic to #829 and #830 once `cxx::set` and `cxx::map` are available.